### DISCUSSION TITLE
Errors on rayimport

### DISCUSSION THREAD

**Discussion by sdhargay** [csiro-robotics/raycloudtools#13] [Q&A] [answered]
Hi Thomas,
I am getting the following error when trying to import a forest pointcloud (~30x30 m plot, 1.5 gb):

$ rayimport pointcloud.ply ray 0,0,0 --max_intensity 0 --remove_start_pos
reading: pointcloud.ply
Warning: no times provided in file, applying 1 second difference per ray consecutively, starting at 0 seconds
warning: no colour information found in pointcloud.ply, setting colours red->green->blue based on time
warning: very large data in point 0, suspicious:      288659 6.27811e+06      27.782
read and process 19 / 21 1.690s
start position:      288659 6.27811e+06      27.782 removed from all points

Next, when I try to run rayextract it seems to run but nothing happens:
$ rayextract terrain pointcloud_raycloud.ply
reading: pointcloud_raycloud.ply
read and process 17 / 21 1.300s

When I tried repeating the above steps using just 
$rayimport pointcloud.ply 0,0,0
it outputs a raycloud file, but when I run rayextract terrain on it, it gives the following error for each point and I have to terminate the process:
....
Error: very large ray length in ray index 61111      -288663 -6.27811e+06     -37.2243, bad input.
Use rayexport then rayimport the exported point cloud with a fixed trajectory file
Error: very large ray length in ray index 61112      -288663 -6.27811e+06     -37.2335, bad input.
Use rayexport then rayimport the exported point cloud with a fixed trajectory file
Error: very large ray length in ray index 61113      -288663 ^Z

I did follow your instructions in an earlier question on the forum, on using ccmake to make DOUBLE_RAYS=ON
![image](https://github.com/csiro-robotics/raycloudtools/assets/1866120/d148434d-f6d3-4b35-ac41-f6a77f9269e3)

But it didn't seem to make a difference. Hope you could suggest a solution.
Thanks for your support !

**Comment by thomas-lowe** [comment, marked as answer]
Hi sdhargay, sorry you are experiencing problems. 

 The warnings for 'rayimport pointcloud.ply 0,0,0' are as expected. It has the sensor origin at 0,0,0 but the cloud many km away, so the rays are too long. I have just fixed the problem of it issuing the warning too much, thanks.
 The warnings for 'rayimport pointcloud.ply ray 0,0,0 --max_intensity 0 --remove_start_pos' are correct but not a problem in your case. Your cloud has no time value or colour value and, although the positions are far from the origin, you are removing the start point.

 The reason rayextract terrain does nothing is that WITH_QHULL is turned off, and rayextract terrain uses QHULL. You can see how to install QHULL in the readme.md. It should have output the error message: "Error: extracting terrain requires QHull, see README instructions for installation"
 
 FYI: rayextract terrain can get very slow on large clouds, as it is superlinear with respect to the number of points. If it takes too long then you can raydecimate and rayextract terrain on the smaller cloud. Functions like rayextract trees don't mind if the ground is based on a decimated cloud, you can still use the full cloud for tree extraction. 
 
p.s. I will add a warning if you use rayimport ray 0,0,0 (as some functions expect the rays to have a length). 

thanks

**Reply by sdhargay** [reply]
Thanks a lot Thomas! 

I enabled WITH_QHULL and followed the instructions given under "Optional build dependencies: For raywrap".

![image](https://github.com/csiro-robotics/raycloudtools/assets/1866120/7ef3ff4c-87cc-4c5d-8421-06f95c6388dd)

Although WITH_QHULL is ON, there are lots of DIRS_NOTFOUND lines as you can see above. I cloned the qhull repo to the same directory-level as my root raycloudtools directory, maybe I need to clone it to the "raycloudtools/build" directory? 

I also tried to install qhull using: $sudo apt install libqhull-dev, which didn't make a difference.

Anyhow I ran $rayextract terrain pointcloud_raycloud.ply" again but there wasn't any output generated although it seemed to do something briefly:
$ rayextract terrain pointcloud_raycloud.ply
reading: pointcloud_raycloud.ply
read and process 18 / 21 1.286s

I tried rayextract terrain again with the "--gradient 1" option, but it was the same outcome.

thanks, sonam

**Reply by thomas-lowe** [reply]
If you followed the instructions including the sudo make install, then the cmake .. should be able to find the library and include paths as long as the install has put it somewhere that is in the standard paths that cmake uses.
This is probably your LD_LIBRARY_PATH, and the place Qhull was installed to is probably /usr/local/lib
You can try running echo $LD_LIBRARY_PATH to see where cmake will look, and if it doesn't include the place it was installed then add to your ~/.bashrc:
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/local/lib
remember to: source ~/.bashrc afterwards, or close and re-open the terminal.

If that doesn't work then let me know where libqhull.so.7.3.2 has been installed to, when you ran the sudo make install.
You may also want to use raycloudtools through docker, where all dependencies are installed for you. See the info in the raycloudtools/docker folder.

**Reply by sdhargay** [reply]
I am running Ubuntu on Windows Sub-system for Linux (WSL). As shown below the libqhull.so.7.3.2 file seems to be present at the expected location: \\\wsl.localhost\Ubuntu\usr\local\lib
![image](https://github.com/csiro-robotics/raycloudtools/assets/1866120/82cc2b98-010b-4b87-a8cb-6ee282a1af4f)

Running echo $LD_LIBRARY_PATH didn't show anything, so I did the "export ..." step you suggested; followed by the "source..."
Then I ran rayextract terrain again, but with the same results as before.

**Reply by thomas-lowe** [reply]
"Then I ran rayextract terrain again"
Did you do the cmake .. -DCMAKE_POSITION_INDEPENDENT_CODE:BOOL=true, then make, then sudo make install?

The Qhull_DIR should be /usr/local/lib/cmake/Qhull if cmake has found it.

If cmake can't find the package and it is pointed to in the LD_LIBRARY_PATH then I won't be any help in debugging it, I'm not an expert in cmake's FIND_PACKAGE systems. You may want to look at the last two run commands in raycloudtools/docker/dockerfile and check that is essentially what you did.

Other that using docker, you could also try using a 3rd party 2D Delauney triangulation in place of the 'hull.growUpwards(0.01)' function in rayterrain.cpp. Then you wouldn't need Qhull at all.

 sorry I couldn't be more help.

**Reply by sdhargay** [reply]
Thank you for your help. Yes I agree it seems to be due to cmake's inability to find the qhull shared library, and could be something specific to my system set-up. I will try to figure it out and post the solution here if I find one.

In the meantime I have followed your suggestion and thankfully "rayextract terrain" and "rayextract trees" is working as expected in Docker!
